pkg/epp/flowcontrol/registry/registry.go

```diff
@@ -17,121 +17,118 @@ limitations under the License.
 package registry
 
 import (
+	"cmp"
 	"context"
 	"fmt"
+	"slices"
 	"sync"
 	"sync/atomic"
+	"time"
 
 	"github.com/go-logr/logr"
 	"k8s.io/utils/clock"
+
 	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/flowcontrol/contracts"
 	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/flowcontrol/framework"
 	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/flowcontrol/types"
 	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/util/logging"
 )
 
+// propagateStatsDeltaFunc defines the callback function used to propagate statistics changes (deltas) up the hierarchy
+// (Queue -> Shard -> Registry).
+// Implementations MUST be non-blocking (relying on atomics).
+type propagateStatsDeltaFunc func(priority uint, lenDelta, byteSizeDelta int64)
+
 // bandStats holds the aggregated atomic statistics for a single priority band across all shards.
 type bandStats struct {
 	byteSize atomic.Int64
 	len      atomic.Int64
 }
 
+// flowState holds all tracking state for a single flow instance within the registry.
+type flowState struct {
+	key types.FlowKey
+
+	// gcLock protects the flow's lifecycle state.
+	// - The Garbage Collector takes an exclusive write lock to safely delete the flow.
+	// - Active connections take a shared read lock for the duration of their operation, preventing the GC from running
+	//   while allowing other connections to proceed concurrently.
+	gcLock sync.RWMutex
+
+	// leaseCount is an atomic reference counter for all concurrent, in-flight connections.
+	// It is the sole source of truth for determining if a flow is Idle.
+	leaseCount atomic.Int64
+
+	// becameIdleAt tracks the time at which the lease count last dropped to zero.
```

Comment:
Clearer to follow and test than the generation marker we used previously... but the same concept.